encodings/fastlanes/src/rle/compress.rs

```diff
@@ -0,0 +1,269 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: Copyright the Vortex contributors
+
+use std::hash::Hash;
+
+use arrayref::{array_mut_ref, array_ref};
+use fastlanes::RLE;
+use vortex_array::arrays::PrimitiveArray;
+use vortex_array::builders::{ArrayBuilder, PrimitiveBuilder};
+use vortex_array::vtable::ValidityHelper;
+use vortex_array::{IntoArray, ToCanonical};
+use vortex_buffer::BufferMut;
+use vortex_dtype::{NativePType, match_each_unsigned_integer_ptype};
+use vortex_error::VortexResult;
+
+use crate::{FL_CHUNK_SIZE, RLEArray};
+
+impl RLEArray {
+    /// Encodes a primitive array of unsigned integers using FastLanes RLE.
+    pub fn encode(array: &PrimitiveArray) -> VortexResult<Self> {
+        vortex_dtype::match_each_unsigned_integer_ptype!(array.ptype(), |T| {
+            rle_encode_typed::<T>(array)
+        })
+    }
+}
+
+/// Decompresses an RLE array back into a primitive array.
+pub fn rle_decompress(array: &RLEArray) -> PrimitiveArray {
+    match_each_unsigned_integer_ptype!(array.ptype(), |T| { rle_decode_typed::<T>(array) })
+}
+
+/// Encodes a primitive array of unsigned integers using FastLanes RLE.
+///
+/// In case the input array length is % 1024 != 0, the last chunk is padded.
+fn rle_encode_typed<T>(array: &PrimitiveArray) -> VortexResult<RLEArray>
+where
+    T: NativePType + RLE + Clone + Hash + Eq,
+{
+    let values = array.as_slice::<T>();
+    let len = values.len();
+
+    // Allocate capacity up to the next multiple of chunk size.
+    let mut values_buf = BufferMut::<T>::with_capacity(len.next_multiple_of(FL_CHUNK_SIZE));
+    let mut indices_buf = BufferMut::<u16>::with_capacity(len.next_multiple_of(FL_CHUNK_SIZE));
+
+    // Pre-allocate for one offset per chunk.
+    let mut value_chunk_offsets = BufferMut::<u64>::with_capacity(len.div_ceil(FL_CHUNK_SIZE));
+
+    let values_uninit = values_buf.spare_capacity_mut();
+    let indices_uninit = indices_buf.spare_capacity_mut();
+    let mut value_count_acc = 0; // Chunk value count prefix sum.
+
+    // Note that the loop iterates the last chunk even if smaller than 1024.
+    for chunk_start_idx in (0..len).step_by(FL_CHUNK_SIZE) {
+        let chunk_end = std::cmp::min(chunk_start_idx + FL_CHUNK_SIZE, len);
+        let chunk_len = chunk_end - chunk_start_idx;
+        let chunk_slice = &values[chunk_start_idx..chunk_end];
+
+        // SAFETY:
+        // `MaybeUninit<T>` and `T` have the same layout.
+        let rle_vals: &mut [T] = unsafe {
+            std::mem::transmute(
+                &mut values_uninit[value_count_acc..value_count_acc + FL_CHUNK_SIZE],
```

Comment:
I think these are nicer expressed as 
```suggestion
                &mut values_uninit[value_count_acc..][..FL_CHUNK_SIZE],
```

Reply:
done